File: tests/wasm/test_uma.js
import test from 'node:test';
import assert from 'node:assert';
import * as wasm from '../../pkg/vector_ta.js';
import {
    loadTestData,
    assertClose
} from './test_utils.js';


let testData;


const EXPECTED_LAST_5 = [
    59665.81830666, 59477.69234542, 59314.50778603,
    59218.23033661, 59143.61473211
];

test('UMA - partial params', () => {
    const data = Array.from({length: 100}, (_, i) => 100.0 + i);

    const result = wasm.uma_js(
        new Float64Array(data),
        1.0,
        5,
        50,
        4,
        null
    );

    assert.ok(result, 'Result should not be null');
    assert.ok(result.values, 'Result should have values array');
    assert.strictEqual(result.values.length, data.length, 'Result length should match input length');
});

test.before(() => {
    testData = loadTestData();
});

test('UMA - accuracy test', () => {
    const result = wasm.uma_js(
        new Float64Array(testData.close),
        1.0,
        5,
        50,
        4,
        null
    );


    const validValues = result.values.filter(v => !isNaN(v));
    const lastValues = validValues.slice(-5);


    for (let i = 0; i < EXPECTED_LAST_5.length && i < lastValues.length; i++) {
        const diff = Math.abs(lastValues[i] - EXPECTED_LAST_5[i]);
        const tolerance = EXPECTED_LAST_5[i] * 0.01;
        assert.ok(
            diff < tolerance || diff < 1.0,
            `Value ${i}: Expected ${EXPECTED_LAST_5[i]}, got ${lastValues[i]}, diff ${diff}`
        );
    }
});

test('UMA - default candles', () => {
    const data = Array.from({length: 100}, (_, i) => 100.0 + i);

    const result = wasm.uma_js(
        new Float64Array(data),
        1.0,
        5,
        50,
        4,
        null
    );

    assert.ok(result, 'Result should not be null');
    assert.strictEqual(result.values.length, data.length, 'Result length should match input length');
});

test('UMA - zero max_length error', () => {
    const data = [10.0, 20.0, 30.0];

    assert.throws(() => {
        wasm.uma_js(
            new Float64Array(data),
            1.0,
            5,
            0,
            4
        );
    }, 'Should throw error for zero max_length');
});

test('UMA - period exceeds length error', () => {
    const data = [10.0, 20.0, 30.0];

    assert.throws(() => {
        wasm.uma_js(
            new Float64Array(data),
            1.0,
            5,
            10,
            4
        );
    }, 'Should throw error when max_length exceeds data length');
});

test('UMA - very small dataset error', () => {
    const data = [42.0];

    assert.throws(() => {
        wasm.uma_js(
            new Float64Array(data),
            1.0,
            5,
            50,
            4
        );
    }, 'Should throw error for insufficient data');
});

test('UMA - empty data error', () => {
    assert.throws(() => {
        wasm.uma_js(
            new Float64Array([]),
            1.0,
            5,
            50,
            4
        );
    }, 'Should throw error for empty data');
});

test('UMA - invalid accelerator error', () => {
    const data = Array.from({length: 100}, (_, i) => 100.0 + i);


    assert.throws(() => {
        wasm.uma_js(
            new Float64Array(data),
            0.5,
            5,
            50,
            4
        );
    }, 'Should throw error for invalid accelerator < 1.0');


    assert.throws(() => {
        wasm.uma_js(
            new Float64Array(data),
            -1.0,
            5,
            50,
            4
        );
    }, 'Should throw error for negative accelerator');
});

test('UMA - invalid min_max error', () => {
    const data = Array.from({length: 100}, (_, i) => 100.0 + i);


    assert.throws(() => {
        wasm.uma_js(
            new Float64Array(data),
            1.0,
            60,
            50,
            4
        );
    }, 'Should throw error when min_length > max_length');
});

test('UMA - NaN handling', () => {
    const data = Array.from({length: 100}, (_, i) => 100.0 + i);

    const result = wasm.uma_js(
        new Float64Array(data),
        1.0,
        5,
        50,
        4,
        null
    );


    const warmupPeriod = 53;
    let nanCount = 0;
    for (let i = 0; i < warmupPeriod && i < result.values.length; i++) {
        if (isNaN(result.values[i])) {
            nanCount++;
        }
    }
    assert.ok(nanCount > 0, 'Should have NaN values in warmup period');


    if (result.values.length > warmupPeriod + 10) {
        const validAfterWarmup = result.values.slice(warmupPeriod + 10).filter(v => !isNaN(v));
        assert.ok(validAfterWarmup.length > 0, 'Should have valid values after warmup');
    }
});

test('UMA - all NaN error', () => {
    const data = new Float64Array(100).fill(NaN);

    assert.throws(() => {
        wasm.uma_js(
            data,
            1.0,
            5,
            50,
            4
        );
    }, 'Should throw error for all NaN data');
});

test('UMA - with leading NaNs', () => {

    const data = new Float64Array(110);
    for (let i = 0; i < 10; i++) {
        data[i] = NaN;
    }
    for (let i = 10; i < 110; i++) {
        data[i] = 100.0 + (i - 10);
    }

    const result = wasm.uma_js(
        data,
        1.0,
        5,
        50,
        4
    );

    assert.strictEqual(result.values.length, data.length, 'Result length should match input');


    const validValues = result.values.slice(70).filter(v => !isNaN(v));
    assert.ok(validValues.length > 0, 'Should handle NaN prefix and produce valid values');
});

test('UMA - with volume data', () => {
    const data = Array.from({length: 100}, (_, i) => 100.0 + i);
    const volume = Array.from({length: 100}, (_, i) => 1000.0 + i * 10);

    const result = wasm.uma_js(
        new Float64Array(data),
        1.0,
        5,
        50,
        4,
        volume
    );

    assert.ok(result, 'Result should not be null');
    assert.ok(result.values, 'Result should have values array');
    assert.strictEqual(result.values.length, data.length, 'Result length should match input length');


    const validValues = result.values.filter(v => !isNaN(v));
    assert.ok(validValues.length > 0, 'Should have valid values after warmup');
});

test('UMA - different parameter combinations', () => {
    const data = Array.from({length: 100}, (_, i) => 100.0 + i * 0.5);


    const result1 = wasm.uma_js(
        new Float64Array(data),
        2.0,
        5,
        50,
        4
    );
    assert.ok(result1, 'Should handle accelerator = 2.0');


    const result2 = wasm.uma_js(
        new Float64Array(data),
        1.0,
        10,
        30,
        4
    );
    assert.ok(result2, 'Should handle different length range');


    const result3 = wasm.uma_js(
        new Float64Array(data),
        1.0,
        5,
        50,
        8
    );
    assert.ok(result3, 'Should handle different smooth_length');
});


test('UMA - zero-copy API', () => {
    const data = testData.close;
    const len = data.length;


    const ptr = wasm.uma_alloc(len);
    assert.ok(ptr !== 0, 'Should allocate memory');


    const view = wasm.uma_get_view(ptr, len);
    for (let i = 0; i < len; i++) {
        view[i] = data[i];
    }


    wasm.uma_update(ptr, len, 1.0, 5, 50, 4);


    const result = wasm.uma_get_view(ptr, len);


    assert.strictEqual(result.length, len, 'Result length should match');


    const validValues = [];
    for (let i = 0; i < result.length; i++) {
        if (!isNaN(result[i])) {
            validValues.push(result[i]);
        }
    }


    if (validValues.length >= 5) {
        const lastValues = validValues.slice(-5);
        for (let i = 0; i < EXPECTED_LAST_5.length && i < lastValues.length; i++) {
            const diff = Math.abs(lastValues[i] - EXPECTED_LAST_5[i]);
            const tolerance = EXPECTED_LAST_5[i] * 0.01;
            assert.ok(
                diff < tolerance || diff < 1.0,
                `Zero-copy value ${i}: Expected ${EXPECTED_LAST_5[i]}, got ${lastValues[i]}`
            );
        }
    }


    wasm.uma_free(ptr, len);
});

test('UMA - zero-copy memory management', () => {
    const sizes = [10, 100, 1000];

    for (const size of sizes) {
        const ptr = wasm.uma_alloc(size);
        assert.ok(ptr !== 0, `Should allocate memory for size ${size}`);

        const view = wasm.uma_get_view(ptr, size);
        assert.strictEqual(view.length, size, `View should have correct size ${size}`);


        for (let i = 0; i < size; i++) {
            view[i] = 100.0 + i;
        }


        wasm.uma_free(ptr, size);
    }
});


test('UMA - batch processing single params', () => {
    const data = Array.from({length: 100}, (_, i) => 100.0 + i);

    const result = wasm.uma_batch_js(
        new Float64Array(data),
        [1.0, 1.0, 0.0],
        [5, 5, 0],
        [50, 50, 0],
        [4, 4, 0]
    );

    assert.ok(result, 'Batch result should not be null');
    assert.ok(result.values, 'Should have values matrix');
    assert.ok(result.accelerators, 'Should have accelerators array');
    assert.ok(result.min_lengths, 'Should have min_lengths array');
    assert.ok(result.max_lengths, 'Should have max_lengths array');
    assert.ok(result.smooth_lengths, 'Should have smooth_lengths array');


    assert.strictEqual(result.rows, 1, 'Should have 1 row');
    assert.strictEqual(result.cols, data.length, 'Should have correct column count');
    assert.strictEqual(result.values.length, data.length, 'Values array should match data length');
});

test('UMA - batch processing multiple params', () => {
    const data = Array.from({length: 50}, (_, i) => 100.0 + i);

    const result = wasm.uma_batch_js(
        new Float64Array(data),
        [1.0, 2.0, 0.5],
        [5, 10, 5],
        [30, 30, 0],
        [4, 4, 0]
    );

    assert.ok(result, 'Batch result should not be null');


    assert.strictEqual(result.rows, 6, 'Should have 6 rows');
    assert.strictEqual(result.cols, data.length, 'Should have correct column count');
    assert.strictEqual(result.values.length, 6 * data.length, 'Values array size should be rows * cols');


    assert.strictEqual(result.accelerators.length, 6, 'Should have 6 accelerator values');
    assert.strictEqual(result.min_lengths.length, 6, 'Should have 6 min_length values');
    assert.strictEqual(result.max_lengths.length, 6, 'Should have 6 max_length values');
    assert.strictEqual(result.smooth_lengths.length, 6, 'Should have 6 smooth_length values');


    for (let row = 0; row < result.rows; row++) {
        const rowStart = row * result.cols;
        const rowEnd = rowStart + result.cols;
        const rowValues = result.values.slice(rowStart, rowEnd);
        const validCount = rowValues.filter(v => !isNaN(v)).length;
        assert.ok(validCount > 0, `Row ${row} should have some valid values`);
    }
});

test('UMA - batch with volume', () => {
    const data = Array.from({length: 50}, (_, i) => 100.0 + i);
    const volume = Array.from({length: 50}, (_, i) => 1000.0 + i * 10);

    const result = wasm.uma_batch_js(
        new Float64Array(data),
        [1.0, 1.0, 0.0],
        [5, 5, 0],
        [30, 30, 0],
        [4, 4, 0],
        volume
    );

    assert.ok(result, 'Batch result with volume should not be null');
    assert.strictEqual(result.rows, 1, 'Should have 1 row');
    assert.strictEqual(result.cols, data.length, 'Should have correct column count');
});


test('UMA - streaming interface', () => {

    const stream = wasm.uma_stream_new(1.0, 5, 50, 4);
    assert.ok(stream, 'Should create stream');


    const results = [];
    for (let i = 0; i < 60; i++) {
        const value = 100.0 + i;
        const result = wasm.uma_stream_update(stream, value);
        results.push(result);
    }


    const nullCount = results.filter(r => r === null || r === undefined || isNaN(r)).length;
    assert.ok(nullCount > 0, 'Should have null/NaN values during warmup');


    const validCount = results.filter(r => r !== null && r !== undefined && !isNaN(r)).length;
    assert.ok(validCount > 0, 'Should have valid values after warmup');


    wasm.uma_stream_reset(stream);


    const afterReset = wasm.uma_stream_update(stream, 100.0);
    assert.ok(afterReset === null || afterReset === undefined || isNaN(afterReset),
        'Should need warmup after reset');


    wasm.uma_stream_free(stream);
});

test('UMA - streaming with volume', () => {

    const stream = wasm.uma_stream_new(1.0, 5, 50, 4);
    assert.ok(stream, 'Should create stream');


    const results = [];
    for (let i = 0; i < 60; i++) {
        const value = 100.0 + i;
        const volume = 1000.0 + i * 10;
        const result = wasm.uma_stream_update_with_volume(stream, value, volume);
        results.push(result);
    }


    const validCount = results.filter(r => r !== null && r !== undefined && !isNaN(r)).length;
    assert.ok(validCount > 0, 'Should have valid values after warmup with volume');


    wasm.uma_stream_free(stream);
});

test('UMA - streaming vs batch comparison', () => {
    const data = testData.close;


    const batchResult = wasm.uma_js(
        new Float64Array(data),
        1.0, 5, 50, 4, null
    );


    const stream = wasm.uma_stream_new(1.0, 5, 50, 4);
    const streamResults = [];

    for (const value of data) {
        const result = wasm.uma_stream_update(stream, value);
        streamResults.push(result !== null && result !== undefined ? result : NaN);
    }


    assert.strictEqual(batchResult.values.length, streamResults.length,
        'Batch and stream should produce same length');



    const batchValid = batchResult.values.filter(v => !isNaN(v));
    const streamValid = streamResults.filter(v => !isNaN(v));

    if (batchValid.length >= 5 && streamValid.length >= 5) {

        const batchLast = batchValid.slice(-5);
        const streamLast = streamValid.slice(-5);

        for (let i = 0; i < batchLast.length && i < streamLast.length; i++) {
            const relativeDiff = Math.abs(batchLast[i] - streamLast[i]) / Math.max(Math.abs(batchLast[i]), 1.0);
            assert.ok(relativeDiff < 0.1,
                `Streaming vs batch mismatch at ${i}: batch=${batchLast[i]}, stream=${streamLast[i]}`);
        }
    }


    wasm.uma_stream_free(stream);
});

console.log('All UMA WASM tests passed!');

File: tests/python/test_uma.py
"""
Python binding tests for UMA indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:

    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS
from rust_comparison import compare_with_rust


class TestUma:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()

    def test_uma_partial_params(self):
        """Test UMA with partial parameters (None values) - mirrors check_uma_partial_params"""
        data = np.arange(100, dtype=np.float64) + 100.0


        result = ta_indicators.uma(data, 1.0, 5, 50, 4, None)
        assert len(result) == len(data)

    def test_uma_accuracy(self, test_data):
        """Test UMA matches expected values from Rust tests - mirrors check_uma_accuracy"""
        close = test_data['close']
        expected = EXPECTED_OUTPUTS['uma']

        result = ta_indicators.uma(
            close,
            accelerator=expected['default_params']['accelerator'],
            min_length=expected['default_params']['min_length'],
            max_length=expected['default_params']['max_length'],
            smooth_length=expected['default_params']['smooth_length'],
            volume=None
        )

        assert len(result) == len(close)


        valid_values = result[~np.isnan(result)]


        assert_close(
            valid_values[-5:] if len(valid_values) >= 5 else valid_values,
            expected['last_5_values'] if len(valid_values) >= 5 else expected['last_5_values'][:len(valid_values)],
            rtol=0.01,
            msg="UMA last 5 values mismatch"
        )

    def test_uma_default_candles(self, test_data):
        """Test UMA with default parameters - mirrors check_uma_default_candles"""
        close = test_data['close']


        result = ta_indicators.uma(close, 1.0, 5, 50, 4, None)
        assert len(result) == len(close)

    def test_uma_zero_max_length(self):
        """Test UMA fails with zero max_length - mirrors check_uma_zero_period"""
        input_data = np.array([10.0, 20.0, 30.0])

        with pytest.raises(ValueError, match="Invalid max_length"):
            ta_indicators.uma(input_data, accelerator=1.0, min_length=5, max_length=0, smooth_length=4, volume=None)

    def test_uma_period_exceeds_length(self):
        """Test UMA fails when max_length exceeds data length - mirrors check_uma_period_exceeds_length"""
        data_small = np.array([10.0, 20.0, 30.0])

        with pytest.raises(ValueError, match="Invalid max_length"):
            ta_indicators.uma(data_small, accelerator=1.0, min_length=5, max_length=10, smooth_length=4, volume=None)

    def test_uma_very_small_dataset(self):
        """Test UMA fails with insufficient data - mirrors check_uma_very_small_dataset"""
        single_point = np.array([42.0])

        with pytest.raises(ValueError, match="Invalid max_length"):
            ta_indicators.uma(single_point, accelerator=1.0, min_length=5, max_length=50, smooth_length=4, volume=None)

    def test_uma_empty_input(self):
        """Test UMA fails with empty input - mirrors check_uma_empty_input"""
        empty = np.array([])

        with pytest.raises(ValueError, match="Input data slice is empty"):
            ta_indicators.uma(empty, accelerator=1.0, min_length=5, max_length=50, smooth_length=4, volume=None)

    def test_uma_invalid_accelerator(self):
        """Test UMA fails with invalid accelerator - mirrors check_uma_invalid_params"""
        data = np.arange(100, dtype=np.float64) + 100.0

        with pytest.raises(ValueError, match="Invalid accelerator"):
            ta_indicators.uma(data, accelerator=0.5, min_length=5, max_length=50, smooth_length=4, volume=None)

        with pytest.raises(ValueError, match="Invalid accelerator"):
            ta_indicators.uma(data, accelerator=-1.0, min_length=5, max_length=50, smooth_length=4, volume=None)

    def test_uma_invalid_min_max(self):
        """Test UMA fails when min_length > max_length"""
        data = np.arange(100, dtype=np.float64) + 100.0

        with pytest.raises(ValueError, match="min_length.*max_length"):
            ta_indicators.uma(data, accelerator=1.0, min_length=60, max_length=50, smooth_length=4, volume=None)

    def test_uma_nan_handling(self, test_data):
        """Test UMA handles NaN values correctly - mirrors check_uma_nan_handling"""
        close = test_data['close']
        expected = EXPECTED_OUTPUTS['uma']

        result = ta_indicators.uma(close, 1.0, 5, 50, 4, None)
        assert len(result) == len(close)


        warmup = expected['warmup_period']
        if len(result) > warmup + 10:

            valid_count = np.sum(~np.isnan(result[warmup + 10:]))
            assert valid_count > 0, "Should have valid values after warmup period"


        nan_count = np.sum(np.isnan(result[:warmup]))
        assert nan_count >= warmup - 1, f"Expected at least {warmup-1} NaN values in warmup period, got {nan_count}"

    def test_uma_streaming(self, test_data):
        """Test UMA streaming matches batch calculation - mirrors check_uma_streaming"""
        close = test_data['close']
        accelerator = 1.0
        min_length = 5
        max_length = 50
        smooth_length = 4


        batch_result = ta_indicators.uma(
            close,
            accelerator=accelerator,
            min_length=min_length,
            max_length=max_length,
            smooth_length=smooth_length,
            volume=None
        )


        stream = ta_indicators.UmaStream(
            accelerator=accelerator,
            min_length=min_length,
            max_length=max_length,
            smooth_length=smooth_length
        )
        stream_values = []

        for price in close:
            result = stream.update(price)
            stream_values.append(result if result is not None else np.nan)

        stream_values = np.array(stream_values)


        assert len(batch_result) == len(stream_values)



        batch_valid = batch_result[~np.isnan(batch_result)]
        stream_valid = stream_values[~np.isnan(stream_values)]

        if len(batch_valid) >= 5 and len(stream_valid) >= 5:

            for i, (b, s) in enumerate(zip(batch_valid[-5:], stream_valid[-5:])):
                relative_diff = abs(b - s) / max(abs(b), 1.0)
                assert relative_diff < 0.1, f"UMA streaming mismatch at index {i}: batch={b}, stream={s}"

    def test_uma_streaming_with_volume(self, test_data):
        """Test UMA streaming with volume data"""
        close = test_data['close']
        volume = test_data['volume']


        batch_result = ta_indicators.uma(
            close,
            accelerator=1.0,
            min_length=5,
            max_length=50,
            smooth_length=4,
            volume=volume
        )


        stream = ta_indicators.UmaStream(
            accelerator=1.0,
            min_length=5,
            max_length=50,
            smooth_length=4
        )
        stream_values = []

        for i, price in enumerate(close):
            vol = volume[i] if i < len(volume) else None
            result = stream.update_with_volume(price, vol)
            stream_values.append(result if result is not None else np.nan)

        stream_values = np.array(stream_values)


        assert len(batch_result) == len(stream_values)


        assert not np.all(np.isnan(stream_values)), "Stream should produce some valid values"

    def test_uma_batch(self, test_data):
        """Test UMA batch processing - mirrors batch test patterns"""
        close = test_data['close']

        result = ta_indicators.uma_batch(
            close,
            accelerator_range=(1.0, 1.0, 0.0),
            min_length_range=(5, 5, 0),
            max_length_range=(50, 50, 0),
            smooth_length_range=(4, 4, 0),
            volume=None
        )

        assert 'values' in result
        assert 'combos' in result
        assert 'rows' in result
        assert 'cols' in result


        assert result['rows'] == 1
        assert result['cols'] == len(close)


        values_2d = result['values']
        if values_2d.ndim == 2:
            default_row = values_2d[0]
        else:

            default_row = values_2d

        expected = EXPECTED_OUTPUTS['uma']


        valid_values = default_row[~np.isnan(default_row)]


        if len(valid_values) >= 5:

            assert_close(
                valid_values[-5:],
                expected['last_5_values'],
                rtol=0.1,
                msg="UMA batch default row mismatch"
            )

    def test_uma_batch_multiple_params(self, test_data):
        """Test UMA batch processing with multiple parameter combinations"""
        close = test_data['close'][:100]

        result = ta_indicators.uma_batch(
            close,
            accelerator_range=(1.0, 2.0, 0.5),
            min_length_range=(5, 10, 5),
            max_length_range=(30, 30, 0),
            smooth_length_range=(4, 4, 0),
            volume=None
        )


        assert result['rows'] == 6
        assert result['cols'] == len(close)


        values_2d = result['values']
        if values_2d.ndim == 2:
            assert values_2d.shape[0] == 6
            assert values_2d.shape[1] == len(close)


            for row in range(6):
                row_values = values_2d[row]
                valid_count = np.sum(~np.isnan(row_values))
                assert valid_count > 0, f"Row {row} should have some valid values"
        else:

            assert len(values_2d) == 6 * len(close)


        assert len(result['combos']) == 6
        for combo in result['combos']:
            assert 'accelerator' in combo
            assert 'min_length' in combo
            assert 'max_length' in combo
            assert 'smooth_length' in combo

    def test_uma_all_nan_input(self):
        """Test UMA with all NaN values"""
        all_nan = np.full(100, np.nan)

        with pytest.raises(ValueError, match="All values are NaN"):
            ta_indicators.uma(all_nan, accelerator=1.0, min_length=5, max_length=50, smooth_length=4, volume=None)

    def test_uma_with_leading_nans(self):
        """Test UMA handles leading NaN values correctly"""

        data = np.concatenate([
            np.full(10, np.nan),
            np.arange(100, dtype=np.float64) + 100.0
        ])

        result = ta_indicators.uma(
            data,
            accelerator=1.0,
            min_length=5,
            max_length=50,
            smooth_length=4,
            volume=None
        )

        assert len(result) == len(data)


        valid_count = np.sum(~np.isnan(result[70:]))
        assert valid_count > 0, "Should handle NaN prefix and produce valid values"

    def test_uma_different_parameters(self):
        """Test UMA with various parameter combinations"""
        data = np.arange(100, dtype=np.float64) + 100.0


        result1 = ta_indicators.uma(data, accelerator=2.0, min_length=5, max_length=50, smooth_length=4, volume=None)
        assert len(result1) == len(data)


        result2 = ta_indicators.uma(data, accelerator=1.0, min_length=10, max_length=30, smooth_length=4, volume=None)
        assert len(result2) == len(data)


        result3 = ta_indicators.uma(data, accelerator=1.0, min_length=5, max_length=50, smooth_length=8, volume=None)
        assert len(result3) == len(data)


        valid1 = result1[~np.isnan(result1)]
        valid2 = result2[~np.isnan(result2)]
        valid3 = result3[~np.isnan(result3)]

        if len(valid1) > 0 and len(valid2) > 0:
            assert not np.allclose(valid1[-1], valid2[-1], rtol=1e-10), "Different parameters should produce different results"

    def test_uma_with_volume(self):
        """Test UMA with volume data"""

        np.random.seed(42)
        data = 100.0 + np.cumsum(np.random.randn(100) * 2)
        volume = 1000.0 + np.random.rand(100) * 1000.0


        result_with_vol = ta_indicators.uma(
            data,
            accelerator=1.0,
            min_length=5,
            max_length=50,
            smooth_length=4,
            volume=volume
        )


        result_no_vol = ta_indicators.uma(
            data,
            accelerator=1.0,
            min_length=5,
            max_length=50,
            smooth_length=4,
            volume=None
        )

        assert len(result_with_vol) == len(data)
        assert len(result_no_vol) == len(data)


        valid_with = result_with_vol[~np.isnan(result_with_vol)]
        valid_without = result_no_vol[~np.isnan(result_no_vol)]

        if len(valid_with) > 0 and len(valid_without) > 0:




            diff = abs(valid_with[-1] - valid_without[-1])

            assert diff > 1e-10, f"Volume should affect UMA calculation (diff={diff})"


if __name__ == '__main__':
    pytest.main([__file__, '-v'])